### ISSUE TITLE
PageController init() Requirements causes login form to error 

### ISSUE THREAD

**Issue by StephenMakrogianni** [silverstripe/silverstripe-login-forms#44]
Hey team,

Including requirements in the PageController init function causes an error when trying to load the login form:
e.g.
```
protected function init()
{
    parent::init();
    Requirements::themedJavascript('dist/app');
}
```
Causes the following:

`[Emergency] Uncaught InvalidArgumentException: The javascript file doesn't exist. Please check if the file dist/app.js exists in any context or search for themedJavascript references calling this file in your templates`

You can work around this by setting the LoginForms login_theme in your config.yml:

```
SilverStripe\LoginForms\EnablerExtension:
  login_themes:
    - app
```

The login form should be completely isolated from any theme or external Controller, so the fix could probably be to stop it from even calling PageController::init().

### Implementation notes
- We'll be looking to proceed along the lines of @Cheddam's recommendations: https://github.com/silverstripe/silverstripe-login-forms/issues/44#issuecomment-538800064
- This issue will focus on resolving it for v4 only.

# Pull requests
* [ ] https://github.com/silverstripe/silverstripe-login-forms/pull/70

**Comment by robbieaverill** [comment]
Hi @StephenMakrogianni, where is this module calling `PageController::init()`?

**Comment by StephenMakrogianni** [comment]
@robbieaverill that was a suggestion from product. I should have clearly marked it as so. I haven't dug deep into it yet.

**Comment by NightJar** [comment]
🤔 
More info required please @StephenMakrogianni 
You have a `themes/app/dist` folder path in your project? Or an `app/dist` folder? If the latter - this is not a theme ;)
Regardless, this is an issue that one would hit if they had a theme enabled, so I'll at least test that flow too before closing.

@robbieaverill This module is responsible for styling the standard login flow, which calls `PageController::init` here: https://github.com/silverstripe/silverstripe-framework/blob/4.4.4/src/Security/Security.php#L586-L588

**Comment by robbieaverill** [comment]
Ok so if there are requirements calls in either a controller or template that you cannot change, referencing files that don’t exist in your project, you can either extend the controller or template in order to remove the offending lines or you can use Requirements::block() to ignore certain calls from being used

**Comment by NightJar** [comment]
For sure there are work-arounds. But I think it should work out of the box.
I think adding `Requirements::clear` to `SilverStripe\LoginForms\EnablerExtension::beforeCallActionHandler`
or
`<% require clear %> to the default template
should prevent this from being an issue.

**Comment by robbieaverill** [comment]
That’s pretty heavy handed, you’ll break any custom frontend plugins by clearing everything

**Comment by NightJar** [comment]
This is true, but remember the point of this module is to 100% replace an existing 'front-end'.

**Comment by robbieaverill** [comment]
Sure, but you’re saying nobody could add their own functionality to this page by doing this. Anyway, as you said it should work out of the box. As far as I can tell it does, so I’m interested to hear what steps are required for this to break in the way described

**Comment by NightJar** [comment]
Requirements clear should call before any custom themes (c.f. `login_themes` config option) are applied (when done in PHP), so customisation would still be easily achievable.

Putting the call in the template would require one to then also have all the other requirements calls in the template, as an impact such as you describe though - but customisation would still be possible.

One potential downside might be a lack of tracking scripts one might have installed via `PageController::init`, I'll admit that. But I'd also revel in not having tracking scripts log keystrokes on a login form. 🤷‍♂

**Comment by robbieaverill** [comment]
Requirements::block() is specific, and works fine for fixing problems like this. Yes, you can use clear() to nuke everything, but I’m not sure why you’d pass up potential edge cases in order to use it when you have a perfect tool for the job already.

**Comment by NightJar** [comment]
Documentation option it is then I guess :)
Requirements::block relies on developers to implement, as there's no way this module can block any theme requirement so far as I know.

**Comment by robbieaverill** [comment]
Requirements calls are global, they don’t care about themes or modules. Developers can block requirements calls from any PHP file (extension, middleware, config.php etc) or from an SS template. The order of execution doesn’t matter, so long as it happens before rendering.

**Comment by NightJar** [comment]
Sorry I wasn't clear.
If a change were to be made to this module in order to block dependencies that cause the issue to manifest, that would require fore-knowledge of all dependencies, OR a way to identify those that have been added as part of a theme (`themedJavascript`) which is no longer a part of of the theme list (removed for favour of `login_themes`).

The former is impossible, and the latter is too tricky - I'm not aware of a way to programmatically determine a requirement's location, other than making assumptions about the path.

This leaves documentation to help make people aware of this limitation, and how to work around it.

**Comment by robbieaverill** [comment]
We are chewing fat anyway, this issue needs more context before it can be resolved properly

**Comment by NightJar** [comment]
> I’m interested to hear what steps are required for this to break in the way described

- Have an existing project which includes a bespoke theme that has JavaScript files
- Those JavaScript files are included via `Requirements` calls from the projects `PageController`
- Update this existing project to use `silverstripe/login-forms` - e.g. by [installing the new `silverstripe/mfa` module](https://github.com/silverstripe/silverstripe-mfa/blob/4.0.1/composer.json#L33)

Issue manifests.

Admittedly it's easy to call this a bit of an edge case, but as we have no way of knowing how a site has been developed - I think it's fair to call this a valid issue that is easily addressed.

**Comment by Cheddam** [comment]
Reading through this, I still feel that stopping the PageController::init() call is the best solve. The intent of this module is to *completely divorce the authentication UI from the front of the website*, in order to make it consistent. A developer should still be able to apply custom CSS / JS if they need to, but anything required by the front shouldn't be applied to the auth UI by default.

I also wouldn't really call this an edge-case in the context of this module:

> - Have an existing project which includes a bespoke theme that has JavaScript files
> - Those JavaScript files are included via Requirements calls from the project's PageController
> - **Update this existing project to use silverstripe/login-forms** - e.g. by installing the new silverstripe/mfa module

There's a strong possibility an implementation of the behaviour of this module will make its way into core in the next major release, likely in the form of dropping the behaviour of using the Page controller by default (or entirely), so I'd rather not muddy the water and force devs to do extra work to adopt the module (especially since it's now required by both MFA and in new silverstripe/installer projects).

Looking at the place where `PageController::init()` is called, we may be able to stop it by wiping out the `Security.page_class` config so it defaults to treating `Security` itself as the response controller.

**Comment by NightJar** [comment]
>stop it by wiping out the Security.page_class config so it defaults to treating Security itself as the response controller

I like this, thanks @Cheddam

**Comment by tardinha** [comment]
Any more thoughts on excluding site theme requirements?

**Comment by NightJar** [comment]
`Security.page_class` is a configuration value, and thus is customisable at a project level as a work around for now, if you should need this before a concrete implementation is released in this module. I can't give a timeline on the latter though, sorry.

Pull requests are always welcome if you would like to give it a go yourself though @tardinha !

**Comment by camfindlay** [comment]
Just come across this myself @Cheddam 

With regards to wiping out `Security.page_class` in config how would this look in the YAML file? A code example here would be helpful to resolve this quickly on projects out in the wild :) Thanks!

**Comment by lozcalver** [comment]
In case it helps anyone else - some of my styles conflicted with the login forms, so I couldn’t use the YAML workaround from the initial ticket. My workaround for now is to wrap all my `Requirements` calls in an if statement:

```php
if ($this->data()->URLSegment !== 'Security') {
    Requirements::themedCSS('my/styles.css');
}
```

**Comment by emteknetnz** [comment]
Fixed in https://github.com/silverstripe/silverstripe-login-forms/pull/70 - implementation is what @Cheddam suggested above

**Comment by emteknetnz** [comment]
The linked PR has been merged, this will be released as part of `4.3.0`, probably during the next quarterly release
